File: src/Interface/MainMenu.hs
module Interface.MainMenu where

import Game.GameLoop (gameLoop)
import GeniusDraw.DrawingFunctions (drawHintGenie, drawMenuGenie, drawStartGenie)
import Interface.StringsAndCommand
import System.Random
import Validate.ValidateMenuInput
  ( validateName,
    validateRange,
  )

initGame :: String -> IO ()
initGame openingSentence = do
  clearT
  drawMenuGenie openingSentence initMenuTitle initMenuOptions
  option <- getLine
  if validateRange 2 option
    then
      if (read option :: Int) == 1
        then mainMenu presentsMainMenu
        else do
          clearT
          drawStartGenie presentsInitialWithdrawal
    else initGame presentsValidEntry

mainMenu :: String -> IO ()
mainMenu openingSentence = do
  clearT
  drawMenuGenie openingSentence mainMenuTitle mainMenuOptions
  option <- getLine
  if validateRange 4 option
    then getMenu (read option)
    else mainMenu presentsValidEntry

getMenu :: Int -> IO ()
getMenu option
  | option == 1 = do
      clearT
      drawMenuGenie presentsOpponents opponentsMenuTitle opponentsMenuOptions
      option <- getLine
      if validateRange 2 option
        then do
          if (read option :: Int) == 1
            then do
              namePlayer1 <- getNamePlayer presentsNamePlayer1
              namePlayer2 <- getNamePlayer presentsNamePlayer2
              list <- createStack 4
              gameLoop list 0 namePlayer1 namePlayer2
            else createDifficulty presentsChallenges
        else getMenu 1
  | option == 2 = do
      clearT
      drawHintGenie presentsChallengesTypes challengeTypes
      x <- getLine
      mainMenu presentsMainChallenge
  | option == 3 = do
      clearT
      drawHintGenie presentsChallengesPerforms performsChallenge
      x <- getLine
      mainMenu presentsMainPerforms
  | option == 4 = drawStartGenie presentsWithdrawal

getNamePlayer :: String -> IO String
getNamePlayer phrase = do
  clearT
  drawStartGenie phrase
  name <- getLine
  if validateName name
    then return name
    else getNamePlayer presentsNameError

createDifficulty :: String -> IO ()
createDifficulty phrase = do
  clearT
  drawMenuGenie phrase challengesMenuTitle challengesMenuOptions
  option2 <- getLine
  if validateRange 3 option2
    then do
      name <- getNamePlayer presentsNamePlayer
      list <- createStack (read option2 :: Int)
      gameLoop list (read option2 :: Int) name ""
    else createDifficulty presentsValidEntry

createStack :: Int -> IO [Int]
createStack difficulty
  | difficulty == 1 = getStacks 10 3
  | difficulty == 2 = getStacks 15 4
  | difficulty == 3 = getStacks 20 5
  | difficulty == 4 = getStacks 15 5

getStacks :: Int -> Int -> IO [Int]
getStacks coins stacks
  | stacks == 1 = return [coins]
  | otherwise = do
      choice <- randomRIO (2, coins `div` stacks)
      restOfList <- getStacks (coins - choice) (stacks - 1)
      return (choice : restOfList)

File: prolog/src/Interface/mainMenu.pl
:- include('../Game/gameLoop.pl').
:- include('../GeniusDraw/DrawingFunctions.pl').
:- include('../Interface/StringsAndCommand.pl').
:- include('../Validate/isValidMenu.pl').

initGame(OpeningSentence) :-
  initMenuTitle(Title),
  initMenuOptions(Options),
  drawMenuGenie(OpeningSentence, Title, Options),
  input(Option2),
  (
    validateRange(2, Option2) ->
      atom_number(Option2, Option),
      redirectInitWindow(Option),
      !
  ; 
    presentsValidEntry(ValidEntryMessage),
    initGame(ValidEntryMessage),
    !
  ).

redirectInitWindow(1):-
  presentsMainMenu(MainMenuMessage),
  mainMenu(MainMenuMessage).

redirectInitWindow(2):-
  presentsInitialWithdrawal(WithdrawalMessage),
  drawStartGenie(WithdrawalMessage),
  halt.

mainMenu(OpeningSentence) :-
  mainMenuTitle(Title),
  mainMenuOptions(Options),
  drawMenuGenie(OpeningSentence, Title, Options),
  input(Option2),
  (
    validateRange(4, Option2) ->
      atom_number(Option2, Option),
      getMenu(Option),
      !
  ; 
    presentsValidEntry(ValidEntryMessage),
    mainMenu(ValidEntryMessage),
    !
  ).

getMenu(1) :-
  call(presentsOpponents, PresentsOp),
  call(opponentsMenuTitle, OpponentsMT),
  call(opponentsMenuOptions, OpponentsMO),
  drawMenuGenie(PresentsOp, OpponentsMT, OpponentsMO),
  input(Option2),
  (
    validateRange(2, Option2) ->
      atom_number(Option2, Option),
      redirectGameMode(Option),
      !
  ; 
    getMenu(1),
    !
  ).

redirectGameMode(1) :-
  presentsNamePlayer1(Player1),
  getNamePlayer(Player1, NamePlayer1),
  presentsNamePlayer2(Player2),
  getNamePlayer(Player2, NamePlayer2),
  createStack(4, Stack),
  gameLoop(Stack, 0, NamePlayer1, NamePlayer2).

redirectGameMode(2) :-
  presentsChallenges(Challenges),
  createDifficulty(Challenges).

getMenu(2) :-
  call(presentsChallengesTypes, PresentsCT),
  call(challengeTypes, ChallengeT),
  drawHintGenie(PresentsCT, ChallengeT),
  input(_),
  call(presentsMainChallenge, Challenges),
  mainMenu(Challenges).

getMenu(3) :-
  call(presentsChallengesPerforms, PresentsCP),
  call(performsChallenge, PerformsC),
  drawHintGenie(PresentsCP, PerformsC),
  input(_),
  call(presentsMainPerforms, Performs),
  mainMenu(Performs).

getMenu(4) :-
  call(presentsWithdrawal, PresentsW),
  drawStartGenie(PresentsW),
  halt.

getNamePlayer(Phrase, Name) :-
  drawStartGenie(Phrase),
  input(Name3),
  atom_string(Name3, Name2),
  (
    validateName(Name2) -> 
      Name = Name2,
      true, 
      !
  ; 
    presentsNameError(NameError),
    getNamePlayer(NameError, Name),
    !
  ).

createDifficulty(Phrase) :-
  challengesMenuTitle(Title),
  challengesMenuOptions(Options),
  drawMenuGenie(Phrase, Title, Options),
  input(Difficulty2),
  (validateRange(3, Difficulty2) ->
    atom_number(Difficulty2, Difficulty),
    call(presentsNamePlayer, PhraseNamePlayer),
    getNamePlayer(PhraseNamePlayer, Name),
    createStack(Difficulty, Stack),
    gameLoop(Stack, Difficulty, Name, ""), !
  ;
    call(presentsValidEntry, PhraseValidEntry),
    createDifficulty(PhraseValidEntry), !
  ),
  !.

createStack(1, Stack) :-
  getStacks(10, 3, Stack).

createStack(2, Stack) :-
  getStacks(15, 4, Stack).

createStack(3, Stack) :-
  getStacks(20, 5, Stack).

createStack(4, Stack) :-
  getStacks(15, 5, Stack).

getStacks(Coins, 1, [Coins]) :- !.

getStacks(Coins, Stacks, R) :-
    Lim is (Coins div Stacks) + 1,
    random(2, Lim, Take),
    RestCoins is Coins - Take,
    RestStacks is Stacks - 1,
    getStacks(RestCoins, RestStacks, R2),
    R = [Take|R2]. 
